src/WebformCivicrmPostProcess.php

```diff
@@ -910,7 +910,19 @@ private function saveGroupsAndTags($contact, $cid, $c) {
         $field_name = 'civicrm_' . $c . '_contact_1_' . $fid;
         if (!empty($contact['other'][1][$type]) || isset($this->enabled[$field_name])) {
           $add = wf_crm_aval($contact, "other:1:$type", array());
+          // ToDo - $add in D7WFC only contains the option(s) selected so unset everything else b/c addOrRemoveMultivaluedData is expecting that
+          foreach ($this->getExposedOptions($field_name) as $k => $v) {
+            if ($add[$k] === 0) {
+              unset($add[$k]);
+            }
+          }
           $remove = empty($this->existing_contacts[$c]) ? array() : $this->getExposedOptions($field_name, $add);
+          // ToDo - $remove should only return the options not selected b/c addOrRemoveMultivaluedData is expecting that
+          foreach ($this->getExposedOptions($field_name) as $k => $v) {
+            if (array_key_exists($k, $add)) {
+              unset($remove[$k]);
+            }
+          }
```

Comment:
seems like this check should be done much before when forming civi data from the webform submission? So that `saveGroupsAndTags()`  already receives the data in the format of what it expects. 

Something like -

```
diff --git a/src/WebformCivicrmBase.php b/src/WebformCivicrmBase.php
index 74b7449..353c978 100644
--- a/src/WebformCivicrmBase.php
+++ b/src/WebformCivicrmBase.php
@@ -530,7 +530,7 @@ abstract class WebformCivicrmBase {
     }
 
     if ($field && $field['#type'] == 'civicrm_options') {
-      return $field['#options'] ?? [];
+      return $field['#options'] ? array_diff_key($field['#options'], $exclude) : [];
     }
     if ($field && $field['#type'] == 'select') {
       // Fetch static options
diff --git a/src/WebformCivicrmPostProcess.php b/src/WebformCivicrmPostProcess.php
index 404667b..bd3f74a 100644
--- a/src/WebformCivicrmPostProcess.php
+++ b/src/WebformCivicrmPostProcess.php
@@ -2394,6 +2394,9 @@ class WebformCivicrmPostProcess extends WebformCivicrmBase implements WebformCiv
               $customValue[] = $itemName;
             }
           }
+          if ($table == 'other' && in_array($name, ['group', 'tag'])) {
+            $val = array_filter($val);
+          }
 
           // Implode data that will be stored as a string
           if ($table !== 'other' && $name !== 'event_id' && $name !== 'relationship_type_id' && $table !== 'contact' && $dataType != 'ContactReference' && strpos($name, 'tag') !== 0) {
```

Can you pls replace and see if it still passes the test added in this PR?

Reply:
👍 Thank you for this @jitendrapurohit.  I'll first expand the test to multiple checkboxes/loading existing data etc first b/c that's working now and needs to be locked in before we further rework this.